File: 2015/05/coins.py
from hashlib import md5

def hash(secret, n):
    return md5(secret + str(n)).hexdigest()

def search(secret, m):
    n = 0
    while True:
        h = hash(secret, n)
        if h.startswith('0'*m):
            return n
        n += 1

#assert(search('abcdef') == 609043)
print 'Part 1:', search('ckczppom', 5)
print 'Part 2:', search('ckczppom', 6)

File: 2015/04/coins.hs
import Data.Digest.Pure.MD5
import qualified Data.ByteString.Lazy.Char8 as L

mhash :: String -> Int -> String
mhash secret n = show . md5 $ L.pack (secret ++ show n)

search' :: String -> Int -> Int -> Int
search' secret len n
    | all (=='0') (take len h) = n
    | otherwise = search' secret len (n+1)
    where h = mhash secret n

search :: String -> Int -> Int
search secret len = search' secret len 0

main :: IO ()
main = do
    putStrLn $ "Part 1: " ++ show (search "ckczppom" 5)
    putStrLn $ "Part 2: " ++ show (search "ckczppom" 6)
